lib/block-supports/layout.php

```diff
@@ -94,52 +92,50 @@ function gutenberg_get_layout_style( $selector, $layout, $has_block_gap_support
 			$justify_content_options += array( 'space-between' => 'space-between' );
 		}
 
-		$flex_wrap_options = array( 'wrap', 'nowrap' );
-		$flex_wrap         = ! empty( $layout['flexWrap'] ) && in_array( $layout['flexWrap'], $flex_wrap_options, true ) ?
-			$layout['flexWrap'] :
-			'wrap';
+		if ( ! empty( $layout['flexWrap'] ) && 'nowrap' === $layout['flexWrap'] ) {
+			$style .= "$selector { flex-wrap: nowrap; }";
+		}
 
-		$style  = "$selector {";
-		$style .= 'display: flex;';
 		if ( $has_block_gap_support ) {
 			if ( is_array( $gap_value ) ) {
 				$gap_row    = isset( $gap_value['top'] ) ? $gap_value['top'] : $fallback_gap_value;
 				$gap_column = isset( $gap_value['left'] ) ? $gap_value['left'] : $fallback_gap_value;
 				$gap_value  = $gap_row === $gap_column ? $gap_row : $gap_row . ' ' . $gap_column;
 			}
-			$gap_style = $gap_value && ! $should_skip_gap_serialization ? $gap_value : "var( --wp--style--block-gap, $fallback_gap_value )";
-			$style    .= "gap: $gap_style;";
-		} else {
-			$style .= "gap: $fallback_gap_value;";
+			if ( $gap_value && ! $should_skip_gap_serialization ) {
```

Comment:
I might have lost some context on this issue. When testing the fallback value for the Columns block, I noticed failing.

Should we be setting `$gap_value` when it's `null`? Something like this?

```suggestion
			if ( $gap_value && ! $should_skip_gap_serialization ) {
				$gap_value = $gap_value ? $gap_value : $fallback_gap_value;
```

Reply:
Oh, thanks @ramonjd, I forgot about that one. This is a tricky part of the problem, and I'm not quite sure what the right solution is. We can't use the fallback value here, or it'll overwrite the root gap value so that if a base `blockGap` value exists then the fallback would take precedence. The objective in this function should be "if there is no gap value at the block level, do not output a rule", and a solution for fallback probably needs to be handled elsewhere, I think?

There's separate handling in this PR for Classic themes and the Columns block fallback value, but I couldn't quite work out a good solution for blocks-based themes and still respecting the fallback value. Let me know if you can think of any other ways of handling it, and I'll do some more 🤔 too!

Reply:
Oh dang, thanks for the explainer. I thought I was missing something. I'll have a think.

Reply:
> The objective in this function should be "if there is no gap value at the block level, do not output a rule",

Does this mean we need to be consistent in the above condition (`is_array( $gap_value )`) and check that at least one vertical value set by the block exists? Otherwise the fallback would take precedence as well. 🤔

Reply:
@andrewserong 

This seems to work for me (so far). Still testing.

<details>

<summary>Diff to check for block.json default values in layout</summary>

```diff
diff --git a/lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php b/lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php
index f8b82319d1..4d4fa857c3 100644
--- a/lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php
+++ b/lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php
@@ -689,15 +689,13 @@ class WP_Theme_JSON_6_1 extends WP_Theme_JSON_6_0 {
 
 		// 4. Generate Layout block gap styles.
 		$has_block_gap_support = _wp_array_get( $this->theme_json, array( 'settings', 'spacing', 'blockGap' ) ) !== null;
-		$has_block_gap_value   = _wp_array_get( $node, array( 'spacing', 'blockGap' ), false );
 
 		if (
 			static::ROOT_BLOCK_SELECTOR !== $selector &&
 			$has_block_gap_support &&
-			$has_block_gap_value &&
 			! empty( $block_metadata['name'] )
 		) {
-			$block_rules .= $this->get_layout_styles( $block_metadata );
+			$block_rules    .= $this->get_layout_styles( $block_metadata );
 		}
 
 		if ( static::ROOT_BLOCK_SELECTOR === $selector ) {
@@ -1164,16 +1162,14 @@ class WP_Theme_JSON_6_1 extends WP_Theme_JSON_6_0 {
 		// Gap styles will only be output if the theme has block gap support, or supports `wp-block-styles`.
 		// In this way, we tie the concept of gap styles to the styles that ship with core blocks.
 		// Default layout gap styles will be skipped for themes that do not explicitly opt-in to blockGap with a `true` or `false` value.
+
 		if ( $has_block_gap_support || $has_block_styles_support ) {
-			$block_gap_value = null;
+			$block_gap_value     = _wp_array_get( $node, array( 'spacing', 'blockGap' ), null );
+			$block_default_value = ! empty( $block_type ) ? _wp_array_get( $block_type->supports, array( 'spacing', 'blockGap', '__experimentalDefault' ), null ) : null;
+			$block_gap_value     = ! $block_gap_value && $block_default_value ? $block_default_value : $block_gap_value;
 			// Use a fallback gap value if block gap support is not available.
 			if ( ! $has_block_gap_support ) {
-				$block_gap_value = '0.5em';
-				if ( ! empty( $block_type ) ) {
-					$block_gap_value = _wp_array_get( $block_type->supports, array( 'spacing', 'blockGap', '__experimentalDefault' ), '0.5em' );
-				}
-			} else {
-				$block_gap_value = _wp_array_get( $node, array( 'spacing', 'blockGap' ), null );
+				$block_gap_value = $block_default_value ? $block_default_value : '0.5em';
 			}
 
 			// Support split row / column values and concatenate to a shorthand value.
diff --git a/lib/compat/wordpress-6.1/theme.json b/lib/compat/wordpress-6.1/theme.json
index 426adc6195..49a142ce01 100644
--- a/lib/compat/wordpress-6.1/theme.json
+++ b/lib/compat/wordpress-6.1/theme.json
@@ -349,6 +349,6 @@
 				}
 			}
 		},
-		"spacing": { "blockGap": "24px" }
+		"spacing": {  }
 	}
 }
diff --git a/test/emptytheme/theme.json b/test/emptytheme/theme.json
index ed2d7b8d09..6817e2939b 100644
--- a/test/emptytheme/theme.json
+++ b/test/emptytheme/theme.json
@@ -7,5 +7,32 @@
 			"wideSize": "1100px"
 		}
 	},
-	"patterns": [ "short-text-surrounded-by-round-images", "partner-logos" ]
+	"styles": {
+		"spacing": {
+			"blockGap": "15px"
+		},
+		"blocks": {
+			"core/columns": {
+				"spacing": {
+					"blockGap": "25px",
+					"padding": "12px"
+				},
+				"color": {
+					"text": "red",
+					"background": "yellow"
+				}
+			},
+			"core/social-links": {
+				"spacing": {
+					"blockGap": "55px",
+					"margin": "100px",
+					"padding": "150px"
+				}
+			}
+		}
+	},
+	"patterns": [
+		"short-text-surrounded-by-round-images",
+		"partner-logos"
+	]
 }
```

</details>

Reply:
Thanks for digging in @ramonjd! That seems to get us a little closer...  but I think with that changeset, unfortunately it depends on there being a `columns` block in the `theme.json` file. I couldn't quite work out a way for us to include the Columns block in a non-hacky way — or perhaps if we need the base `theme.json` file to include the Columns block somehow? (Though it wouldn't have any rules in it 🤔)

Reply:
> perhaps if we need the base theme.json file to include the Columns block somehow

Yeah, maybe we need need to merge these default values into global styles where they don't exist?

That sounds like a whole lotta looping though. 😄

Reply:
Wait, is there a Columns-specific fallback gap value? Where is it stored?

Reply:
> Wait, is there a Columns-specific fallback gap value? Where is it stored?

https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/columns/block.json#L32

Reply:
> Yes! Since https://github.com/WordPress/gutenberg/pull/41098

Thanks, I was just about to link to that. I think storing it in `block.json` made good sense when we were definitely handling the layout support at render time for each individual block. Now that we're centralising things in global styles, I'm not sure if that approach still holds up, so we could potentially revert that change if we can think of a better way of handling it.

It sure is a tricky one, but the main reason we need it is to ensure backwards compatibility for the Columns gap 🤔

Reply:
Would adding the following to https://github.com/WordPress/gutenberg/blob/trunk/lib/compat/wordpress-6.1/theme.json cut it for now?

```json
		"blocks": {
			"core/columns": {
				"spacing": {
					"blockGap": "4em"
				}
			}
		},
```

Reply:
> Would adding the following to https://github.com/WordPress/gutenberg/blob/trunk/lib/compat/wordpress-6.1/theme.json cut it for now?

Unfortunately I don't think so, because then the value would override the root `styles.spacing.blockGap` value for blocks-based themes. I believe we only want the fallback value to hit if there's no blockGap values in use anywhere.

Reply:
Oh I see! Wasn't that fallback aimed at classic themes though? Fwiw I can still see the 2em value applied when testing on classic themes, both with block styles enabled and not.

Reply:
If we're happy with the fallback just existing for Classic themes, then we're probably okay. I think it's probably less of a problem for blocks-based themes since there's a clear path to adding the desired gap value, which doesn't exist for classic themes?

Reply:
I was more thinking that for block themes we already have a kind of fallback in the gap value provided by `.is-layout-flex`, but yeah, it's easy to set a value globally too.

Reply:
> I think it's probably less of a problem for blocks-based themes since there's a clear path to adding the desired gap value, which doesn't exist for classic themes?

That sounds legit. Columns are used a lot for layouts so it might have a significant influence on block sites though until they add a value.

Reply:
> so it might have a significant influence on block sites though until they add a value.

That's a good point, particularly given that Block themes still need to explicitly opt-in to blockGap support (either directly or via `appearanceTools`). So long as they do that, though, the root `blockGap` will be in use. I wonder how many blocks-based themes _don't_ opt-in to block gap at all?

Either way, this is a good opportunity to rethink how the default / fallback value _should_ work. I just haven't (yet) managed to think of an approach that doesn't feel super hacky! 😄

Reply:
@ramonjd and @tellthemachines, I think I've come up with a way to handle the fallback gap in a way that doesn't feel _too_ hacky. In the Resolver class, where individual blocks are iterated over, I've added a section that puts in a `null` placeholder for the `blockGap` value for that particular block, if the block provides a default value. This way, the block is added to the list of blocks to be iterated over by the Theme JSON class.

Then, as currently, over in the Theme JSON class, when we reach the block, we can check if there is blockGap support, and if there isn't, and the current block provides its own default blockGap value, then we use it. My thinking about this not being _too_ hacky is that 1) the Resolver is still responsible for determining whether or not there is a value, and 2) the Theme JSON class is still responsible for determining _which_ value to output.

Seems to work so far (added in this commit: https://github.com/WordPress/gutenberg/pull/40875/commits/422f6654393f0e6085a9e0655eb35c1c6cc22238)  for both blocks-based and classic themes that opt out of block support, but let me know if can foresee any issues with the approach. (And thanks Ramon, it's mostly based on your idea in your diff above, just with a bit extra for the Resolver class 🙇).

Here's a screenshot of a blocks-based theme's site frontend, that has opted out of blockGap support and so receives the fallback `2em` Columns gap, with Social Icons using the base default `0.5em` gap:

![image](https://user-images.githubusercontent.com/14988353/177908544-3f6ede1e-055d-4da4-89f7-492792d073d4.png)

Reply:
Looks like I've broken the JS tests, I'll look into fixing up the logic there shortly 😅

Reply:
Should be fixed now 🤞 as of https://github.com/WordPress/gutenberg/pull/40875/commits/b3f8ba82435879fb87ca38a6ccb21a9f04ab44c4

Reply:
Sounds good to me! Testing by removing `"appearanceTools": true` from the theme's `theme.json`, I can confirm that the default `2em` gap set in the Columns block is working.

One question (not strictly related to this changeset): what's the purpose of the `0.5em` gap value set on the `body` element? It seems too small a value to ever be really useful as a gap.

Reply:
Thanks for re-testing!

> what's the purpose of the `0.5em` gap value set on the body element

That's the one that gets attached to `body .is-layout-flex`, so it's the root default gap that is used by Buttons and Social Icons blocks (or any of the blocks that use the Flex layout). They don't need an explicit `__experimentalDefault` since they're happy using the common base fallback gap. In this PR, rather than always outputting that `0.5em` value at render time of the individual block, we use this root rule so we can avoid duplication.

I liked your idea in one of the other comments of a follow-up PR to remove the tie to the `body` element for generating the base layout styles, though, it might make the relationship here a little clearer 🙂

Reply:
Right, looking at the other blocks with flex layout, `0.5em` seems to work for them. I'd thought Navigation might be an exception, but since nav items already have a fair bit of padding around them, it works well 😄 

> (in effect, we're currently using the body element as a "please only output this rule once and at the root" piece of logic, which could be abstracted in a clearer way, potentially)

Yeah, we're using it to add global rules that don't need to be attached to any particular block/HTML element. Good thing to look at in a follow-up!

Reply:
> I think I've come up with a way to handle the fallback gap in a way that doesn't feel too hacky. In the Resolver class, where individual blocks are iterated over, I've added a section that puts in a null placeholder for the blockGap value for that particular block, if the block provides a default value. 

Thanks! Nice work.

It took me a while to realize that I had to remove blockGap block support from theme.json (even though I don't know why?) After that it tests very well!

I think my assumptions must be wrong on this issue: isn't the point of the columns default fallback to provide a fallback if there is no blockGap value set anywhere, regardless of whether a theme has opted into `spacing.blockGap` or not? 

If not, and I'm way off, then it seems like a good approach!

Reply:
Thanks for taking another look at this! 🙇

> isn't the point of the columns default fallback to provide a fallback if there is no blockGap value set anywhere, regardless of whether a theme has opted into spacing.blockGap or not

It's a good question of what the desired behaviour is here, but in principle the behaviour in this PR should now match `trunk`. Because Gutenberg/core's `theme.json` provides a base `spacing.blockGap` value, the fallback gap will never be reached unless gap support is completely switched off. My current thinking is that if a theme uses gap support (which is opt-in), then it is the theme that's responsible for determining its own Columns spacing, rather than relying on a fallback value.

Or to put it slightly differently, the columns default fallback value is for backwards compatibility, and as of this PR, with the ability to set gap at the block-level in theme.json, if a theme uses `blockGap` then (ideally 😅) it should use the appropriate `core/columns` block in `styles.blocks` rather than depending on fallback styles, if the theme needs to have a particular value for the gap.

It's totally worth re-examining those assumptions if they're incorrect, though! I suppose my preference would be to see if a) the mechanism in this PR is viable, so that b) we feel comfortable enough that we can tweak the behaviour in follow-up PRs if need be 🙂

Reply:
> It's a good question of what the desired behaviour is here, but in principle the behaviour in this PR should now match trunk. Because Gutenberg/core's theme.json provides a base spacing.blockGap value, the fallback gap will never be reached unless gap support is completely switched off. My current thinking is that if a theme uses gap support (which is opt-in), then it is the theme that's responsible for determining its own Columns spacing, rather than relying on a fallback value.

Ah thanks for setting me straight (again 😄) 

That makes total sense.